File: cocro-bff/src/main/kotlin/com/cocro/application/grid/dto/GridDto.kt
package com.cocro.application.grid.dto

import com.cocro.kernel.grid.enums.CellType
import com.cocro.kernel.grid.enums.ClueDirection
import com.cocro.kernel.grid.enums.GridDifficulty
import com.cocro.kernel.grid.enums.SeparatorType

interface GridDto {
    val gridId: String?
    val title: String?
    val difficulty: GridDifficulty?
    val reference: String?
    val description: String?
    val width: Int?
    val height: Int?
    val cells: List<CellDto>?
}

data class SubmitGridDto(
    override val gridId: String? = null,
    override val title: String,
    override val difficulty: GridDifficulty,
    override val reference: String?,
    override val description: String?,
    override val width: Int,
    override val height: Int,
    override val cells: List<CellDto>,
) : GridDto

data class PatchGridDto(
    override val gridId: String,
    override val title: String?,
    override val difficulty: GridDifficulty?,
    override val reference: String?,
    override val description: String?,
    override val width: Int?,
    override val height: Int?,
    override val cells: List<CellDto>?,
) : GridDto

data class ClueDto(
    val direction: ClueDirection,
    val text: String,
)

data class CellDto(
    val x: Int,
    val y: Int,
    val type: CellType,
    val letter: String?,
    val separator: SeparatorType?,
    val number: Int?,
    val clues: List<ClueDto>?,
)

File: cocro-angular/src/app/application/dto/grid.dto.ts
import { Cell, CellType, ClueDirection, GridDifficulty, SeparatorType } from '@domain/models/grid.model';

export interface ClueDto {
  direction: ClueDirection;
  text: string;
}

export interface CellDto {
  x: number;
  y: number;
  type: CellType;
  letter?: string;
  separator?: SeparatorType;
  number?: number;
  clues?: ClueDto[];
}

export interface SubmitGridRequest {
  gridId?: string;
  title: string;
  difficulty: GridDifficulty;
  reference?: string;
  description?: string;
  width: number;
  height: number;
  cells: CellDto[];
  globalClueLabel?: string;
  globalClueWordLengths?: number[];
}

export interface PatchGridRequest {
  gridId: string;
  title?: string;
  difficulty?: GridDifficulty;
  reference?: string;
  description?: string;
  width?: number;
  height?: number;
  cells?: CellDto[];
  globalClueLabel?: string;
  globalClueWordLengths?: number[];
}

export interface GridSubmitResponse {
  gridId: string;
}

export function cellToDto(cell: Cell): CellDto {
  return {
    x: cell.x,
    y: cell.y,
    type: cell.type,
    letter: cell.letter?.value || undefined,
    separator: cell.letter?.separator === 'NONE' ? undefined : cell.letter?.separator,
    number: cell.letter?.number,
    clues: cell.clues?.length ? cell.clues : undefined,
  };
}
